libp2p/autotls/peeridauth.nim

```diff
@@ -0,0 +1,202 @@
+
+proc peerIdSign(
+    privateKey: PrivateKey,
+    challenge: string,
+    publicKey: PublicKey,
+    hostname: string,
+    clientSender: bool = true,
+): string =
+  let prefix = "libp2p-PeerID"
+  let parts =
+    # parts need to be sorted alphabetically by key
+    if clientSender:
+      @[
+        SigParam(k: "challenge-client", v: challenge.toByteSeq()),
+        SigParam(k: "hostname", v: hostname.toByteSeq()),
+        SigParam(k: "server-public-key", v: publicKey.getBytes().get()),
+      ]
+    else:
+      @[
+        SigParam(k: "challenge-server", v: challenge.toByteSeq()),
+        SigParam(k: "client-public-key", v: publicKey.getBytes().get()),
+        SigParam(k: "hostname", v: hostname.toByteSeq),
+      ]
+  let bytesToSign = genDataToSign(prefix, parts)
+  return base64.encode(privateKey.sign(bytesToSign).get().getBytes(), safe = true)
+
+proc checkSignature(
+    serverSig: string,
+    serverPublicKey: PublicKey,
+    challengeServer: string,
+    clientPublicKey: PublicKey,
+    hostname: string,
+): bool {.raises: [PeerIDAuthError].} =
+  let prefix = "libp2p-PeerID"
+  let parts: seq[SigParam] =
+    @[
+      SigParam(k: "challenge-server", v: challengeServer.toByteSeq()),
+      SigParam(k: "client-public-key", v: clientPublicKey.getBytes().get()),
+      SigParam(k: "hostname", v: hostname.toByteSeq),
+    ]
+  let signedBytes = genDataToSign(prefix, parts)
+  var serverSignature: Signature
+  try:
+    if not serverSignature.init(base64.decode(serverSig).toByteSeq):
+      raise newException(
+        PeerIDAuthError, "Failed to initialize Signature from base64 encoded sig"
+      )
+  except ValueError:
+    raise newException(PeerIDAuthError, "Failed to decode server's signature")
+  return serverSignature.verify(
+    signedBytes.toOpenArray(0, signedBytes.len - 1), serverPublicKey
+  )
+
+proc peerIdAuthenticate(
+    url: string, session: HttpSessionRef, peerInfo: PeerInfo, payload: JsonNode
+): Future[(string, HttpClientResponseRef)] {.
+    async: (raises: [AutoTLSError, PeerIDAuthError, CancelledError])
+.} =
+  # Authenticate in three ways as per the PeerID Auth spec
+  # https://github.com/libp2p/specs/blob/master/http/peer-id-auth.md
+
+  # request authentication
+  let base36PeerId = encodePeerId(peerInfo.peerId)
+  var authStartResponse: HttpClientResponseRef
+  try:
+    authStartResponse = await HttpClientRequestRef.get(session, url).get().send()
+  except HttpError:
+    raise newException(PeerIDAuthError, "Failed to start PeerID Auth")
+
+  # www-authenticate
+  let wwwAuthenticate = authStartResponse.headers.getString("www-authenticate")
+  if wwwAuthenticate == "":
+    raise newException(PeerIDAuthError, "www-authenticate not present in response")
+  let challengeClient = extractField(wwwAuthenticate, "challenge-client")
+  var serverPublicKey: PublicKey
+  try:
+    serverPublicKey = PublicKey
+      .init(decode(extractField(wwwAuthenticate, "public-key")).toByteSeq)
+      .get()
```

Comment:
Could `init` fail? I think the call to `get()` assumes it does not (not sure what exception it raises tho).